[feature] 픽 목록 편집 모드

meanjoo opened the pull request:
## #️⃣ 연관된 이슈
- resolved: #54 

## 📝 작업 내용 및 코드
픽 목록에서 top app bar actions의 편집 아이콘을 누르면 다중 삭제가 가능하도록 구현했습니다.

https://github.com/user-attachments/assets/e2ef8c81-03a8-4b87-891f-4e8eecf2d494

<h3><a href="https://vaulted-system-3ae.notion.site/161f85098cd580b590c6fea19a723847?pvs=74">구현 과정 보기</a></h3>

meanjoo committed:
[refactor] 함수 분리

meanjoo committed:
[feature] 픽 목록 편집 모드 활성화/비활성화 버튼 추가

meanjoo committed:
[feature] 편집 모드일 때 항목 끝에 체크 아이콘 표시

meanjoo committed:
Merge branch 'develop' into edit-pick-list

meanjoo committed:
[refactor] DeletePickDialog를 공통 컴포넌트로 수정

meanjoo committed:
[feature] 편집 모드에서 항목 선택 및 해제

meanjoo committed:
[feature] 선택된 항목 삭제하기

meanjoo committed:
[fix] 잘못 설정된 string resource 수정

meanjoo committed:
[refactor] 편집 모드일 때는 선택된 개수 표시

meanjoo committed:
[fix] 편집 모드 하단 버튼 리컴포지션

편집 모드에서 항목을 클릭할 때마다
취소 버튼도 리컴포지션되는 문제 해결

yuni-ju commented on app/src/main/java/com/squirtles/musicroad/common/AlertStringDialog.kt:
```diff
@@ -22,18 +23,18 @@ import androidx.compose.ui.text.font.FontWeight
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
-fun DeletePickDialog(
+internal fun AlertStringDialog(
```
공통으로 사용할 수 있는 Dialog로 바꾼 것 너무 좋습니다!
다만 AlertStringDialog 이름에서 String의 의도가 불분명한 것 같습니다.
MessageAlertDialog이나 ContentAlertDialog 등의 이름으로 변경하는 것은 어떤지 제안드려봅니다..! 

그리고 컴포즈 컨벤션에 따라 파라미터에서 람다를 아래쪽에 배치하면 좋을 것 같습니다!

meanjoo replied:
오옷 MessageAlertDialog 좋은 것 같네요!!! 이 이름으로 반영해두겠습니다
AlertDialog라고 짓자니 deprecated 된 것 같긴 하지만 Compose에 존재하는 AlertDialog와 겹쳐서 다른 이름으로 짓고 싶었는데 마땅한 게 떠오르지 않아 body가 String이라서 AlertStringDialog라고 지었었습니다 ㅎㅎ..

어쩜 제가 고민했던 게 딱 코멘트로 달리는지🙇‍♀️ 좋은 의견 감사합니다

아래 코멘트에 첨부해주신 컴포즈의 파라미터 순서 컨벤션을 봤는데 `@Composable` 람다가 가장 마지막 순서 아닌가요?? `() -> Unit`는 필수 파라미터인 일반 람다라 생각하여 상단에 둔 거였습니다!
title이나 body보다 onDismissRequest가 더 상단에 있는 이유는 .. 큰 의미 없이 함수 안에서 사용한 BasicAlertDialog의 가장 첫 파라미터가 onDismissRequest여서 제가 만드는 함수의 첫 파라미터도 onDismissRequest로 둔 것 같네요
![image](https://github.com/user-attachments/assets/edd3105c-b37a-4344-b2c9-6cc1d2db425c)

onDismissRequest를 body와 buttons 사이로 옮기는 게 더 나을까요?

yuni-ju replied:
앗 그러네요! 말씀해주신대로 일반 람다는 컨벤션의 4번 조건에 해당하지 않는 것 같습니다. ㅎㅎ 그대로 둬도 괜찮을 것 같습니다

yuni-ju commented on app/src/main/java/com/squirtles/musicroad/common/PickInfoText.kt:
```diff
@@ -71,6 +71,7 @@ fun CommentText(
 @Composable
 fun TotalCountText(
     modifier: Modifier = Modifier,
+    prefixText: String = stringResource(R.string.total_count_text),
     totalCount: Int,
     defaultColor: Color = MaterialTheme.colorScheme.onSurface,
     pointColor: Color = MaterialTheme.colorScheme.primary,
```
<img width="440" alt="image" src="https://github.com/user-attachments/assets/6c4d335d-8bb2-49ef-a5f9-8ab11c6b66d6" />

여기도 필수 파라미터인 totalCount가 앞으로 가야할 것 같습니다. 

그리고 개인적으로 TotalCountText의 함수 명에 대해 
TotalCountText면 어차피 prefix가 전체만 있는 것이 아닌가?라고 생각했다가, "선택"도 사용하는 것을 보았습니다!
하지만 좀 더 포괄적인 이름으로 바꾸면 좋을 것 같기도하고, 그대로 둬도 상관없을 것 같습니다.
그냥 이런 생각이 들었다 정도로 읽어주세요! ㅎㅎ

meanjoo replied:
앗 그러네요
TotalCountText의 파라미터 순서를 totalCount, modifier, prefixText, defaultColor, pointColor, style로 변경하겠습니다.

흠 듣고 보니 컴포넌트명이나 prefix 얘도 이름이 변경되어도 좋을 것 같네요! 선택지를 여러 개 준비해봤는데 하나 골라주시면 파라미터 순서와 함께 반영하겠습니다! 저는 개인적으로 4번입니닷ㅎ
1. 함수명 유지(TotalCountText), prefixText 유지 => 현상태로 유지
2. 함수명 유지(TotalCountText), prefixText -> countLabel
3. 함수명 TotalCountText -> CountText, prefixText 유지
4. 함수명 TotalCountText -> CountText, prefixText -> countLabel

+) 이 커밋 푸시할 때 EditModeButton 파라미터 순서도 함께 할게요!

yuni-ju replied:
저도 4번 좋은 것 같습니다! 😄

meanjoo replied:
반영했습니다 😀

yuni-ju commented on app/src/main/java/com/squirtles/musicroad/picklist/EditModeBottomButton.kt:
```diff
@@ -0,0 +1,88 @@
+}
+
+@Composable
+private fun EditModeButton(
+    modifier: Modifier,
+    onClick: () -> Unit,
+    text: String,
+    textColor: Color = White,
+    buttonColor: Color = Primary,
```
여기도 파라미터 순서 확인 부탁드려요!

yuni-ju reviewed (changes requested):
async, await 활용해서 처리해주신 내용과 리컴포지션 최소화 작업 잘 읽었습니다! 
디테일한 부분까지 신경써서 개발해주시는 것에 감사합니다 👍👍

meanjoo committed:
[refactor] 함수명 변경: AlertStringDialog -> MessageAlertDialog

meanjoo committed:
[refactor] 컴포저블 함수명 및 파라미터 순서 변경

TotalCountText -> CountText
prefixText -> countLabel
컨벤션에 따라 CountText, EditModeButton 파라미터 순서 변경

miller198 commented on app/src/main/res/values/strings.xml:
```diff
@@ -77,8 +84,16 @@
     <string name="delete_pick_dialog_delete">삭제하기</string>
 
+    <string name="default_alert_dialog_title">안내</string>
+    <string name="no_selected_pick_dialog_body">삭제 선택한 픽이 없습니다.</string>
```
'삭제 선택한 픽이 없습니다'
보단 그냥 '선택한 픽이 없습니다' 라고 하는게 좋을 것 같습니다!

miller198 commented on app/src/main/java/com/squirtles/musicroad/picklist/DeleteSelectedPickDialog.kt:
```diff
@@ -0,0 +1,56 @@
+        onDismissRequest = onDismissRequest,
+        title = stringResource(if (selectedPickCount == 0) R.string.default_alert_dialog_title else R.string.delete_pick_dialog_title),
+        body = if (selectedPickCount == 0) {
+            stringResource(R.string.no_selected_pick_dialog_body)
```
여기서 선택한 픽이 없을때는 다이얼로그를 띄우는 것보다
삭제하기 버튼을 비활성화 하는게 UX 측면에서 낫지않을까요?

meanjoo replied:
오 넵 그게 더 좋을 것 같네요!

miller198 commented on app/src/main/java/com/squirtles/musicroad/picklist/PickListScreen.kt:
```diff
@@ -200,6 +162,125 @@ fun PickListScreen(
+        }
+
+        if (showDeletePickDialog) {
+            DeleteSelectedPickDialog(
```
EditModeBottomButton와 DeleteSelectedPickDialog 가 PickList 안에서 처리되는게 맞을까요?PickList와는 의미적으로 관계가 없는 것 같네요. 밖으로 빼도 동일하게 동작이 가능할 것 같은데 어떻게 생각하시나요?

meanjoo replied:
앗 그러게요

EditModeBottomButton와 DeleteSelectedPickDialog의 처리를 PickList에서 하기보다는 PickListScreen에서 하는 게 의미상 더 맞을 것 같네요
이 둘을 PickListScreen으로 옮기는 방향으로 수정해보겠습니다!

meanjoo replied:
피드백 반영하여 커밋했습니다!

혹시 버튼이 비활성화 상태일 때 색은 지금 상태가 괜찮을까요?
현재 상태는 아래와 같습니다. 취소 버튼은 DarkGray, 선택 삭제 버튼은 비활성화 되었을 때 Gray입니다.

|선택한 픽이 없을 때|선택한 픽이 있을 때|
|:---:|:---:|
|<img src="https://github.com/user-attachments/assets/d5c3cb3f-baa6-4532-b9a3-ccb19cdaea1f" width="300">|<img src="https://github.com/user-attachments/assets/7afdde44-8b2f-410d-9c5c-fb8b417c6d79" width="300">|

기존에는 취소 버튼이 Gray, 선택 삭제는 Primary였는데 이 상태에서 비활성화 상태의 선택 삭제 버튼 색상도 Gray로 설정하니 취소 버튼과의 구분이 안 되어서 마치 취소 버튼도 비활성화된 버튼인가? 라고 생각되어 다른 색으로 해야겠다고 생각했습니다.
![image](https://github.com/user-attachments/assets/49245bf9-6eb1-4e00-9229-ed42d3276642)

취소 버튼은 Gray색으로 유지하고, 비활성화된 선택 삭제 버튼을 DarkGray로 설정한 건 아래와 같은데 제 눈에는 뭔가 취소 버튼이 더 비활성화 상태인 버튼 같아서 현재(취소 버튼 DarkGray, 비활성화된 선택 삭제 버튼 Gray)처럼 해뒀습니다.

<img src="https://github.com/user-attachments/assets/c5162e7d-aff1-4ec7-a53f-23da0362211a" width="300">

확인하시고 버튼 색상과 관련하여 더 좋은 의견 있으면 반영하겠습니다!

miller198 reviewed (commented):
늦어서 죄송합니다 🙏
진행과정을 상세히 적어주셔서 이해하기 정말 편했어요! 감사합니다
common 패키지에 재활용이 용이한 UI로 바꾸어 만들어주신것도 정말 좋네요

meanjoo committed:
[refactor] 피드백 반영

선택한 픽이 없으면 선택 삭제 버튼을 비활성화
PickList 성격에 맞지 않는 처리가 PickListScreen에서 수행되도록

meanjoo committed:
[refactor] 비활성화 버튼 색상 변경

miller198 reviewed (approved):
👍👌🙌
